File: app/test/visibility_test.dart
// Which customers, leads and routes a rep may see.
//
// Checked against `shared/fixtures/visibility.json`, the same file the
// dashboard's `client/src/domain/__tests__/visibility.test.ts` reads. This one
// decides whether one unit's customers can be seen by another, so it is exactly
// the rule that must not differ between the phone and the dashboard.

import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import 'package:manna_field_sales/core/visibility.dart';

Map<String, dynamic> _fixture() {
  final f = File('../shared/fixtures/visibility.json');
  return json.decode(f.readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  final fx = _fixture();

  final people = (fx['people'] as List)
      .map((e) => VisPerson.fromRow((e as Map).cast<String, dynamic>()))
      .toList();

  group('shared fixture: the field names and the pooled units', () {
    test('reads the fields the fixture names', () {
      final f = (fx['fields'] as Map).cast<String, dynamic>();
      expect(kFieldUnit, f['unit']);
      expect(kFieldCustomerOwner, f['customer_owner']);
      expect(kFieldLeadOwner, f['lead_owner']);
      expect(kFieldRouteOwner, f['route_owner']);
    });

    test('pools exactly the units the fixture pools', () {
      expect(kPooledUnits, fx['pooled_units']);
    });
  });

  group('shared fixture: which units are pooled', () {
    for (final raw in fx['pooled'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        expect(isPooledUnit(c['unit'] as String?), c['expect']);
      });
    }
  });

  group('shared fixture: who sees whose records', () {
    for (final raw in fx['visible'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        final got = visibleReps(people, c['person'] as String?)..sort();
        if (c['expect'] != null) {
          final want = (c['expect'] as List).map((e) => '$e').toList()..sort();
          expect(got, want);
        }
        for (final x in (c['expect_excludes'] as List? ?? const [])) {
          expect(got, isNot(contains('$x')));
        }
      });
    }
  });

  group('shared fixture: what an unassigned record needs to be seen', () {
    for (final raw in fx['owner_values'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        final got = visibleOwnerValues(people, c['person'] as String?)..sort();
        final want = (c['expect'] as List).map((e) => '$e').toList()..sort();
        expect(got, want);
      });
    }
  });

  group('shared fixture: who may assign an owner', () {
    for (final raw in fx['can_assign'] as List) {
      final c = (raw as Map).cast<String, dynamic>();
      test(c['why'] as String, () {
        expect(canAssignOwner(people, c['person'] as String?), c['expect']);
      });
    }
  });

  // --------------------------------------------------- phone-side detail ---

  group('pooling never leaks across units', () {
    test('shows a UAE rep no Indian customers, and vice versa', () {
      final uae = visibleReps(people, 'Kailas Babu');
      for (final other in ['Amjad Pr', 'Sirajudheen Kasim', 'Prasad V']) {
        expect(uae, isNot(contains(other)));
      }
      expect(visibleReps(people, 'Amjad Pr'), ['Amjad Pr']);
    });

    test('never returns the whole company for anybody', () {
      for (final p in people) {
        expect(visibleReps(people, p.name).length, lessThan(people.length));
      }
    });
  });

  group('it fails closed', () {
    test('gives an unresolvable login nothing rather than everything', () {
      expect(visibleReps(people, null), isEmpty);
      expect(visibleReps(people, 'nobody@nowhere'), isEmpty);
      expect(visibleReps(const [], 'Kailas Babu'), isEmpty);
    });

    test('does not read a missing unit as membership of the pool', () {
      final orphan = [
        ...people,
        const VisPerson(name: 'No Unit', unit: ''),
      ];
      expect(visibleReps(orphan, 'No Unit'), ['No Unit']);
    });

    test('leaves a disabled person with nothing, rather than their old team',
        () {
      expect(visibleReps(people, 'Ex UAE Rep'), isEmpty);
    });
  });

  group('who is covering for whom', () {
    test('knows a UAE rep is looking at a colleague\'s customer', () {
      expect(isCoveringFor(people, 'Kailas Babu', 'Manikandan'), isTrue);
    });

    test('does not call a rep\'s own customer a cover', () {
      expect(isCoveringFor(people, 'Kailas Babu', 'Kailas Babu'), isFalse);
      expect(isCoveringFor(people, 'Kailas Babu', null), isFalse);
    });

    test('is never true outside a pooled unit, because nothing is shared', () {
      expect(isCoveringFor(people, 'Amjad Pr', 'Sirajudheen Kasim'), isFalse);
    });

    test('says a UAE rep shares and an Indian rep does not', () {
      expect(sharesWithUnit(people, 'Renjith'), isTrue);
      expect(sharesWithUnit(people, 'Amjad Pr'), isFalse);
      expect(sharesWithUnit(people, 'nobody'), isFalse);
    });
  });
}

File: client/src/domain/__tests__/visibility.test.ts
/**
 * Which customers, leads and routes a rep may see.
 *
 * Checked against `shared/fixtures/visibility.json`, the same file
 * `app/test/visibility_test.dart` reads. This one decides whether one unit's
 * customers can be seen by another, so it is exactly the rule that must not
 * differ between the phone and the dashboard.
 */

import { describe, expect, it } from 'vitest';
import cases from '../../../../shared/fixtures/visibility.json';
import {
  POOLED_UNITS,
  VISIBILITY_FIELD,
  canAssignOwner,
  isCoveringFor,
  isPooledUnit,
  sharesWithUnit,
  visibleOwnerValues,
  visibleReps,
  type Person,
} from '../visibility';

/** The fixture speaks raw ERPNext field names; the domain speaks its own. */
const PEOPLE: Person[] = cases.people.map((p) => ({
  name: p.name,
  unit: p.custom_company,
  enabled: p.enabled === 1,
  isGroup: p.is_group === 1,
}));

describe('shared fixture: the field names and the pooled units', () => {
  it('reads the fields the fixture names', () => {
    expect(VISIBILITY_FIELD.unit).toBe(cases.fields.unit);
    expect(VISIBILITY_FIELD.customerOwner).toBe(cases.fields.customer_owner);
    expect(VISIBILITY_FIELD.leadOwner).toBe(cases.fields.lead_owner);
    expect(VISIBILITY_FIELD.routeOwner).toBe(cases.fields.route_owner);
  });

  it('pools exactly the units the fixture pools', () => {
    expect([...POOLED_UNITS]).toEqual(cases.pooled_units);
  });
});

describe('shared fixture: which units are pooled', () => {
  for (const c of cases.pooled) {
    it(c.why, () => {
      expect(isPooledUnit(c.unit)).toBe(c.expect);
    });
  }
});

describe('shared fixture: who sees whose records', () => {
  for (const c of cases.visible) {
    it(c.why, () => {
      const got = visibleReps(PEOPLE, c.person);
      if (c.expect) expect(got.sort()).toEqual([...c.expect].sort());
      for (const excluded of c.expect_excludes ?? []) expect(got).not.toContain(excluded);
    });
  }
});

describe('shared fixture: what an unassigned record needs to be seen', () => {
  for (const c of cases.owner_values) {
    it(c.why, () => {
      expect(visibleOwnerValues(PEOPLE, c.person).sort()).toEqual([...c.expect].sort());
    });
  }
});

describe('shared fixture: who may assign an owner', () => {
  for (const c of cases.can_assign) {
    it(c.why, () => {
      expect(canAssignOwner(PEOPLE, c.person)).toBe(c.expect);
    });
  }
});

// ------------------------------------------------- dashboard-side detail ---

describe('pooling never leaks across units', () => {
  it('shows a UAE rep no Indian customers, and an Indian rep no UAE ones', () => {
    const uae = visibleReps(PEOPLE, 'Kailas Babu');
    for (const other of ['Amjad Pr', 'Sirajudheen Kasim', 'Prasad V']) {
      expect(uae).not.toContain(other);
    }
    expect(visibleReps(PEOPLE, 'Amjad Pr')).toEqual(['Amjad Pr']);
  });

  it('never returns the whole company for anybody', () => {
    for (const p of PEOPLE) {
      expect(visibleReps(PEOPLE, p.name).length).toBeLessThan(PEOPLE.length);
    }
  });
});

describe('it fails closed', () => {
  it('gives an unresolvable login nothing rather than everything', () => {
    // An empty list is handed to the API as "show nothing". A filter of
    // `undefined` would be read as no filter, and answer with every customer
    // in the company — the bug that was live on 12 Aug 2026.
    expect(visibleReps(PEOPLE, undefined)).toEqual([]);
    expect(visibleReps(PEOPLE, 'nobody@nowhere')).toEqual([]);
    expect(visibleReps([], 'Kailas Babu')).toEqual([]);
  });

  it('does not read a missing unit as membership of the pool', () => {
    const orphan: Person[] = [{ name: 'No Unit', unit: '', enabled: true, isGroup: false }];
    expect(visibleReps([...PEOPLE, ...orphan], 'No Unit')).toEqual(['No Unit']);
  });

  it('leaves a disabled person with nothing, rather than their old team', () => {
    expect(visibleReps(PEOPLE, 'Ex UAE Rep')).toEqual([]);
  });
});

describe('who is covering for whom', () => {
  it('knows a UAE rep is looking at a colleague’s customer', () => {
    expect(isCoveringFor(PEOPLE, 'Kailas Babu', 'Manikandan')).toBe(true);
  });

  it('does not call a rep’s own customer a cover', () => {
    expect(isCoveringFor(PEOPLE, 'Kailas Babu', 'Kailas Babu')).toBe(false);
    expect(isCoveringFor(PEOPLE, 'Kailas Babu', undefined)).toBe(false);
  });

  it('is never true outside a pooled unit, because nothing is shared there', () => {
    expect(isCoveringFor(PEOPLE, 'Amjad Pr', 'Sirajudheen Kasim')).toBe(false);
  });

  it('says a UAE rep shares and an Indian rep does not', () => {
    expect(sharesWithUnit(PEOPLE, 'Renjith')).toBe(true);
    expect(sharesWithUnit(PEOPLE, 'Amjad Pr')).toBe(false);
    expect(sharesWithUnit(PEOPLE, 'nobody')).toBe(false);
  });
});
